text/0000-cargo-binary-dependencies.md

```diff
@@ -0,0 +1,144 @@
+
+```toml
+[build-dependencies]
+cmake = { version = "1.0", type = "bin" }
+```
+
+Cargo will build the `cmake` binary, then make it available to your `build.rs` through an environment variable:
+
+```rust
+// build.rs
+use std::{env, process::Command};
+
+fn main() {
+    let cmake_path = env::var_os("CARGO_BIN_FILE_CMAKE_cmake").expect("cmake binary");
+    let mut cmake = Command::new(cmake_path).arg("--version");
+    assert!(cmake.status().expect("cmake --version failed").success());
+}
+```
+
+You can optionally specify specific binaries to depend on using `bins`:
+
+```toml
+[build-dependencies]
+cmake = { version = "1.0", type = "bin", bins = ["cmake"] }
+```
+
+If no binaries are specified, all the binaries in the package will be built and made available.
+
+You can obtain the directory containing all binaries built by the `cmake` crate with `CARGO_BIN_DIR_CMAKE`, such as to add it to `$PATH` before invoking another build system or a script.
+
+Cargo also allows depending on `cdylib` or `staticlib` artifacts. For example, you can embed a dynamic library in your binary:
+
+```rust
+// main.rs
+const MY_PRELOAD_LIB: &[u8] = include_bytes!(env!("CARGO_CDYLIB_FILE_MYPRELOAD"));
+```
+
+Note that cargo cannot help you ensure these artifacts are available at runtime for an installed version of a binary; cargo can only supply these artifacts at build time. Runtime requirements for installed crates are out of scope for this change.
+
+If you need to depend on multiple variants of a crate, such as both the binary and library of a crate, you can supply an array of strings for `type`: `type = ["bin", "lib"]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are four `type`s available:
+1. `"lib"`, the default
+2. `"bin"`, a crate building one or more binaries
+3. `"cdylib"`, a C-compatible dynamic library
+4. `"staticlib"`, a C-compatible static library
+
+`"lib"` corresponds to all crates that can be depended on currently,
+including `lib`, `rlib`, and `proc-macro` libraries.
+See [linkage](https://doc.rust-lang.org/reference/linkage.html) for more information.
+
+Artifact dependencies can appear in any of the three sections of dependencies (or in target-specific versions of these sections):
+- `[build-dependencies]`
+- `[dependencies]`
+- `[dev-dependencies]`
+
+By default, `build-dependencies` are built for the host, while  `dependencies` and `dev-dependencies` are built for the target. You can specify the `target` attribute to build for a specific target, such as `target = "wasm32-wasi"`; a literal `target = "target"` will build for the target even if specifing a build dependency. (If the target is not available, this will result in an error at build time, just as if building the specified crate with a `--target` option for an unavailable target.)
```

Comment:
One worry I would have about supplying a `target` feature is that this sort of attribute is often better placed on the package itself. For example if you're depending on a wasm blob the package that compiles to wasm is unlikely to ever *not* want to get compiled to wasm, but this means that you'd have to specify the target every single time you depend on it.

I think that this sort of attribute on a dependency would work but it isn't clearly the best solution available to us I think. I think it would be worthwhile to explore the use cases of this and see if they'd be better suited with a dependency-level or package-level target declaration.

Reply:
Are you suggesting something like `[package] default-target = "wasm32-wasi"`? I'd be interested in that.

Reply:
> One worry I would have about supplying a target feature is that this sort of attribute is often better placed on the package itself. For example if you're depending on a wasm blob the package that compiles to wasm is unlikely to ever not want to get compiled to wasm, but this means that you'd have to specify the target every single time you depend on it.

I disagree: for power-cpu-sim (linked above), it is specifically designed to be able to compile to a command-line program for linux/etc. and the exact same crate is also designed to be compiled for the web using target wasm32-wasi and there is an example web server which runs on linux and requires the produced wasm file to run.

Reply:
To be clear, I'm not saying that a dependency-level directive isn't useful. Nor does one reason why it is useful mean it's more useful than a package-level directive. My point is that the package-level directive is probably useful in more cases.

@jyn514 yes that's what I'm thinking (subject to bikeshedding of course)

Reply:
@alexcrichton A package-level directive for the default also makes sense, but that would be a separate proposal. If a package truly supports only *one* target, a package-level directive would suffice. The `target` attribute on a dependency is intended for cases where the dependency supports multiple targets.

For example, consider a crate that builds virtual machine firmware, and supports multiple architectures. If you're building a virtual machine for a specific architecture, you need firmware for that architecture.

Or, consider a kernel that supports both 64-bit and 32-bit userspace applications, and needs to build different versions of a VDSO for each case.

I also expect `target = "target"` on build dependencies to be quite common: "no, I don't want to *run* this as part of the build process, I want this to be runnable on the target system instead".

I would also expect to see cases like this:

```toml
[build-dependencies]
thing32 = { package = "thing", version = "...", target = "i686-unknown-linux-gnu" }
thing64 = { package = "thing", version = "...", target = "x86_64-unknown-linux-gnu" }
```